### PR TITLE
Add protobuf and dashapi to sonic-mgmt

### PR DESCRIPTION
**Pull request by Pterosaur**
#### Why I did it
The testcases in sonic-mgmt need the packages of protobuf and dashapi

##### Work item tracking
- Microsoft ADO **(number only)**:

#### How I did it
Because the docker of sonic-mgmt is based on ubuntu20.04, it cannot directly install the packages compiled by slave due to dependency issues. Download related packaged directly from Azp.

#### How to verify it
Check azp stats.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211
- [ ] 202305

#### Tested branch (Please provide the tested image version)

- [ ] 
- [ ]

### CODE REVIEW

**Comment by lguohan** [comment]
i think the software should be installed in docker-ptf not docker-mgmt

**Comment by lguohan** [comment]
where do we plan to run the sonic-mgmt pytest? is it running from sonic-mgmt or is it from docker-ptf?

**Comment by lguohan** [comment]
can you use docker_sonic_mgmt_debs, so that people if they want to build locally, they can still build it without depending on azp?

**Comment by Pterosaur** [comment]
> where do we plan to run the sonic-mgmt pytest? is it running from sonic-mgmt or is it from docker-ptf?

All protobuf and dash payloads are for control plane which is loaded in sonic-mgmt, docker-ptf is only used pack-gen of dataplane. So, To insteall  protobuf and dash in sonic-mgmt is reasonable.

**Comment by Pterosaur** [comment]
> can you use docker_sonic_mgmt_debs, so that people if they want to build locally, they can still build it without depending on azp?

I tried it before. But the docker of sonic-mgmt is based on ubuntu20.04 instead of debian bullseye, There are some differences on dependencies.
For example, the python3.9 is for bullseye but python3.8 is for ubuntu20.04. All `docker_sonic_mgmt_debs` are compiled under bullseye, so when I want to install the necessary package, python3-protobbuf.deb, it will be broken due to 3.8 < 3.9.
If I forcefully upgrade the python3.8 to python3.9, there are many existing dependencies are broken again,
Python version is one but not ONLY one dependency issue. So, the easiest way is to download them from Azp.

But in the future, when the sonic-mgmt is upgraded to ubuntu22.04, I believe we can directly install them from `docker_sonic_mgmt_debs`.

**Comment by lguohan** [comment]
Let's just build and install in sonic-mgmt container?

**Comment by lguohan** [comment]
Let's just build and install in sonic-mgmt container?

**Review by lguohan — APPROVED** [review]

File: dockers/docker-sonic-mgmt/Dockerfile.j2 [hunk 1]
```diff
@@ -147,6 +148,27 @@ RUN dpkg -i \
 debs/{{ deb }}{{' '}}
 {%- endfor %}
 
+# Install protobuf 3.21.12
+RUN mkdir -p /tmp/protobuf \
+    cd /tmp/protobuf \
+    && wget https://sonicstorage.blob.core.windows.net/public/ubuntu/20.04/libprotobuf-dev_3.21.12-3_amd64.deb \
```
Comment (qiluo-msft) [inline dockers/docker-sonic-mgmt/Dockerfile.j2:154]: >20.04 [](http://example.com/codeflow?start=69&length=5)

It is also used in FROM in the beginning of the file. Will dpkg fail if we upgrade the FROM line to new version?

File: dockers/docker-sonic-mgmt/Dockerfile.j2 [hunk 2]
```diff
@@ -147,6 +148,27 @@ RUN dpkg -i \
...
+    && wget https://sonicstorage.blob.core.windows.net/public/ubuntu/20.04/libprotoc32_3.21.12-3_amd64.deb \
+    && wget https://sonicstorage.blob.core.windows.net/public/ubuntu/20.04/protobuf-compiler_3.21.12-3_amd64.deb \
+    && wget https://sonicstorage.blob.core.windows.net/public/ubuntu/20.04/python3-protobuf_3.21.12-3_amd64.deb \
+    && dpkg -i *.deb \
+    && rm -rf /tmp/protobuf
+
+# Install dash-api
+RUN cd /tmp \
+    && mkdir -p /usr/lib/python3/dist-packages/dash_api \
+    && wget https://raw.githubusercontent.com/sonic-net/sonic-buildimage/master/src/sonic-dash-api/pypkg/__init__.py -O /usr/lib/python3/dist-packages/dash_api/__init__.py \
```
Comment (qiluo-msft) [inline dockers/docker-sonic-mgmt/Dockerfile.j2:167]: >__init__.py [](http://example.com/codeflow?start=105&length=11)

This file is inside the same repo, why wget?

File: dockers/docker-sonic-mgmt/Dockerfile.j2 [hunk 3]
```diff
@@ -147,6 +148,27 @@ RUN dpkg -i \
...
+    && git clone https://github.com/sonic-net/sonic-dash-api.git \
```
[the remark this answers is not in the document]
Comment (lguohan) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:168]: this is within the docker, not available.

[the remark this answers is not in the document]
Comment (qiluo-msft) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:168]: Do we need to worry if this code on a release branch, but it actually wget/git clone some master files?

File: dockers/docker-sonic-mgmt/Dockerfile.j2 (hunk 1 above)
Comment (Pterosaur) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:154]: These packets are compiled in ubuntu20.04. Actually, a better method is to install these packets from bullseye slave by sonic_mgmt_debs variable, but due to bullseye and ubuntu dependencies issues, like python 3.8 and python 3.9, we have to install a pre-compiled version.
If we upgrade this docker to a new version, like ubuntu22.04, we can follow the typical method to build this docker.

File: dockers/docker-sonic-mgmt/Dockerfile.j2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pterosaur) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:168]: 1. If we want to access the files in src, we have to mount the src folder to Docker build workspace. But the files in src may be too many for Docker daemon which will dramatically downgrade the performance of Docker daemon.
2. After the sonic-mgmt upgrade to 20.04, I believe we can follow the typical method, docker_mgmt_debs, to install them.

File: dockers/docker-sonic-mgmt/Dockerfile.j2 [hunk 4]
```diff
@@ -36,7 +36,8 @@ RUN apt-get update && apt-get install -y apt-transport-https \
                                          sudo \
                                          tcpdump \
                                          telnet \
-                                         vim
+                                         vim \
+                                         unzip
```
Comment (xumia) [inline dockers/docker-sonic-mgmt/Dockerfile.j2:40]: The unzip is no use in the latest commit, it can be removed, right?

Comment (Pterosaur) [reply dockers/docker-sonic-mgmt/Dockerfile.j2]: Done

File: dockers/docker-sonic-mgmt/Dockerfile.j2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:167]: Please explore # 1: you do not need to mount src folder, just copy what you need to Docker build workspace.

[the remark this answers is not in the document]
Comment (Pterosaur) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:167]: Under the current framework, I didn't find a script to do that.

**Review by lguohan — APPROVED** [review]

File: dockers/docker-sonic-mgmt/Dockerfile.j2 (hunk 3 above)
[the remark this answers is not in the document]
Comment (qiluo-msft) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:167]: I created a issue https://github.com/sonic-net/sonic-buildimage/issues/15848. And ask for solving in future PR.

File: dockers/docker-sonic-mgmt/Dockerfile.j2 (hunk 1 above)
Comment (qiluo-msft) [reply dockers/docker-sonic-mgmt/Dockerfile.j2:153]: Please consider 20.04 as a named constant.

**Pull request merged**